File: oangiapy/poker/poker-main/app/PokerV2/Hand/Hand.php
<?php

namespace App\PokerV2\Hand;

use App\PokerV2\Card;

/**
  * Hand
  * 
  * @package    Poker
  * @author     oangia <hqnhatdn@gmail.com>
  *
  */

class Hand
{
    const       ZITCH               = 1;
    const       ONEPAIR             = 2;
    const       TWOPAIR             = 3;
    const       THREEKIND           = 4;
    const       STRAIGHT            = 5;
    const       FLUSH               = 6;
    const       FULLHOUSE           = 7;
    const       FOURKIND            = 8;
    const       STRAIGHTFLUSH       = 9;

    public      $point              = 0;
    public      $zitchPoint         = 0;
    public      $cards              = [];
    public      $chiAt              = false;

    function __construct(array $cards, $chiAt = false)
    {
        $this->cards = $cards;
        $this->chiAt = $chiAt;
        $this->pointCalc();
    }

    public function compare(Hand $hand, $zitch = false)
    {
        if ($this->level < $hand->level) return -1;
        if ($this->level > $hand->level) return 1;
        if ($this->level == $hand->level) {
            if ($this->point > $hand->point){
                return 1;
            }
            if ($this->point < $hand->point) {
                return -1;
            }
            if ($this->point == $hand->point) {
                return ($zitch) ? $this->compareZitchPoint($hand) : 0;
            }
        }
    }

    public function compareZitchPoint(Hand $hand)
    {
        if ($this->zitchPoint > $hand->zitchPoint) {
            return 1;
        }
        if ($this->zitchPoint < $hand->zitchPoint) {
            return -1;
        }
        if ($this->zitchPoint == $hand->zitchPoint) {
            return 0;
        }
    }

    public function duplicate(Hand $hand)
    {
        foreach ($this->cards as $card) {
            foreach ($hand->cards as $card2) {
                if ($card->name == $card2->name) {
                    return true;
                }
            }
        }
        return false;
    }

    public function duplicateCard(Card $card)
    {
        foreach ($this->cards as $item) {
            if ($item->name == $card->name) {
                return true;
            }
        }
        return false;
    }

    public function toString()
    {
        $str = '';
        foreach ($this->cards as $card) {
            $str .= $card->name . ',';
        }
        return trim($str, ',');
    }

    public function visualize() {
        $str = '';
        foreach ($this->cards as $card) {
            $str .= $card->visualize();
        }
        return $str;
    }
}

File: oangiapy/poker/core/Hand.py
from oangiapy.poker.core.HandDetector import HandDetector

class Hand:
    ZITCH = 0
    ONEPAIR = 1
    TWOPAIR = 2
    THREEKIND = 3
    STRAIGHT = 4
    FLUSH = 5
    FULLHOUSE = 6
    FOURKIND = 7
    STRAIGHTFLUSH = 8
    
    def __init__(self, cards):
        self._cards = cards
        self._detector = HandDetector

    def get_cards(self):
        return self._cards

    def get_type(self):
        return self._detector.get_type()

    def get_point(self):
        return self._detector.get_point()

    def get_type_str(self):
        names = ["Zitch", "One Pair", "Two Pairs", "Three of a Kind", "Straight", "Flush", "Full House", "Four of a Kind", "Straight Flush"]
        return names[self.get_type()] 

    def check_dup(self, other):
        return any(c1.get_name() == c2.get_name() for c1 in self.get_cards() for c2 in other.get_cards())

    def compare(self, other, zitch=False):
        if self.get_type() < other.get_type():
            return -1
        if self.get_type() > other.get_type():
            return 1
        # same level → compare point
        if self.get_point() > other.get_point():
            return 1
        if self.get_point() < other.get_point():
            return -1
            
        return 0
        
    def __lt__(self, other):
        return self.compare(other) == -1

    def __gt__(self, other):
        return self.compare(other) == 1

    def __repr__(self):
        return ",".join(c.get_name() for c in self._cards) + " " + self.get_type_str() + " " + str(self.get_point())    
